# Profitable App Profiles for the App Store and Google Play Markets

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means our revenue for any given app is mostly influenced by the number of users who use our app — the more users that see and engage with the adds, the better. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.

In [2]:
from csv import reader

In [3]:
opened_file = open("AppleStore.csv")
read_file = reader(opened_file)
ios_dataset = list(read_file)
ios_header = ios_dataset[0]
ios_dataset = ios_dataset[1:]

In [4]:
opened_file = open("googleplaystore.csv")
read_file = reader(opened_file)
android_dataset = list(read_file)
android_header = android_dataset[0]
android_dataset = android_dataset[1:]

In [5]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [6]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [7]:
explore_data(android_dataset, 0, 4, rows_and_columns = True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


In [8]:
print(len(ios_header))
print("\n")
print(ios_header)

16


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


The Apple Store dataset has 16 columns. For this analysis, we can use the following columns from the dataset: "id", "track_name", "price", "rating_count_tot", "user_rating", "price_genre"

In [9]:
print(len(android_header))
print("\n")
print(android_header)

13


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The Google dataset has 13 columns in total. For this analysis, we can use the following columns from the dataset: "App", "Category", "Rating", "Reviews", "Price", "Genres"

Deleting Wrong Data 

In [10]:
print(android_header)
print("\n")
print(android_dataset[10472])
print(len(android_dataset[10472]))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
12


Row 10472 of the Google dataset has an error. The rating column is missing. We can delete this row

In [11]:
del android_dataset[10472]

In [12]:
print(android_dataset[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


In [13]:
for rows in android_dataset:
    name = rows[0]
    if name == "Instagram":
        print(rows)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [14]:
duplicate_apps = []
unique_apps = []

for rows in android_dataset:
    name = rows[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

In [15]:
print(duplicate_apps)
print(len(duplicate_apps))
print(len(unique_apps))

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents', 'Quick PDF Scanner + OCR FREE', 'Genius Scan - PDF Scanner', 'Tiny Scanner - PDF Scanner App', 'Fast Scanner : Free PDF Scan', 'Mobile Doc Scanner (MDScan) Lite', 'TurboScan: scan documents and receipts in PDF', 'Tiny Scanner Pro: PDF Doc Scan', 'Docs To Go™ Free Office Suite', 'OfficeSuite : Free Office + PDF Editor',

The Google dataset has 1181 duplicate apps and 9659 unique apps. From looking at the Instagram duplicates, we can see that the higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app.

In [16]:
reviews_max = {}

for rows in android_dataset:
    name = rows[0]
    n_reviews = float(rows[3])
    if (name in reviews_max) and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

In [17]:
print(len(reviews_max))

9659


In [18]:
android_clean = []
already_added = []

for rows in android_dataset:
    name = rows[0]
    n_reviews = float(rows[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        android_clean.append(rows)
        already_added.append(name)

In [19]:
print(len(android_clean))

9659


Removing Non_English Apps

We're not interested in keeping these kind of apps, so we'll remove them. One way to go about this is to remove each app whose name contains a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.).

In [20]:
def are_english_apps(string):
    for chars in string:
        if ord(chars) > 127:
            return False
    return True

In [21]:
print(are_english_apps('Instachat 😜'))
print(are_english_apps('Docs To Go™ Free Office Suite'))
print(are_english_apps('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(are_english_apps("Instagram"))

False
False
False
True


we wrote a function that detects non-English app names, but we saw that the function couldn't identify correctly certain English app names like 'Docs To Go™ Free Office Suite' and 'Instachat 😜'. This is because emojis and some characters like ™ fall outside the ASCII range and have corresponding numbers that are over 127.

If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emoji or other special characters will still be labeled as English.

In [22]:
def are_english_apps(string):
    non_ascii = 0
    
    for chars in string:
        if ord(chars) > 127:
            non_ascii += 1
    if non_ascii > 3:
        return False
    else:
        return True
        

In [23]:
print(are_english_apps("Docs To Go™ Free Office Suite"))
print(are_english_apps("Instachat 😜"))
print(are_english_apps("爱奇艺PPS -《欢乐颂2》电视剧热播"))

True
True
False


In [24]:
ios_english = []
android_english = []

for rows in android_clean:
    name = rows[0]
    if are_english_apps(name):
        android_english.append(rows)
        
for rows in ios_dataset:
    name = rows[1]
    
    if are_english_apps(name):
        ios_english.append(rows)

In [25]:
print(len(android_english))
print(len(ios_english))

9614
6183


Isolating Free Apps

we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis.

In [26]:
print(ios_english[:4])

[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']]


In [27]:
ios_free_apps = []

for rows in ios_english:
    price = rows[4]
    
    if price == "0.0":
        ios_free_apps.append(rows)

In [28]:
print(len(ios_free_apps))

3222


In [29]:
print(android_english[:3])

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


In [30]:
android_free = []

for rows in android_english:
    price = rows[7]
    
    if price == "0":
        android_free.append(rows)

In [31]:
print(len(android_free))

8864


Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

In [32]:
print(android_header)
print("\n")
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


For the android dataset, we can use the "Category", "Rating" and "Genres" columns to see which apps are more profitable. For the IOS dataet, we can use "prime_genre" and "user_rating" columns

In [33]:
def freq_table(dataset, index):
    freq_dictionary = {}
    total = 0
    
    for rows in dataset:
        value = rows[index]
        total += 1
        
        if value in freq_dictionary:
            freq_dictionary[value] += 1
        else:
            freq_dictionary[value] = 1
    
    percentage_dictionary = {}
    
    for key in freq_dictionary:
        percentage = (freq_dictionary[key]/total) * 100
        percentage_dictionary[key] = percentage
    return percentage_dictionary

In [34]:
freq_table(ios_free_apps, 10)

{'12+': 17.070142768466788,
 '17+': 6.424581005586592,
 '4+': 65.85971446306642,
 '9+': 10.645561762880199}

In [35]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [36]:
print(display_table(android_free, 1))
print("\n")
print(display_table(android_free, 9))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [37]:
print(display_table(ios_free_apps, 11))

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665
None


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

In [38]:
print(display_table(android_free, 1))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [39]:
print(display_table(android_free, 9))

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

There are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

This showed us that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now, we'd like to get an idea about the kind of apps with the most users.

In [40]:
unique_ios_genres = freq_table(ios_free_apps, 11)

for genre in unique_ios_genres:
    total = 0
    len_genre = 0
    
    for rows in ios_free_apps:
        genre_app = rows[11]
        if genre_app == genre:
            rating = float(rows[5])
            total += rating
            len_genre += 1
    avg_rating = total/len_genre
    print(genre, ":", avg_rating)

Productivity : 21028.410714285714
Weather : 52279.892857142855
Lifestyle : 16485.764705882353
Navigation : 86090.33333333333
Business : 7491.117647058823
Entertainment : 14029.830708661417
Education : 7003.983050847458
Finance : 31467.944444444445
News : 21248.023255813954
Games : 22788.6696905016
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Medical : 612.0
Book : 39758.5
Utilities : 18684.456790123455
Food & Drink : 33333.92307692308
Social Networking : 71548.34905660378
Reference : 74942.11111111111
Catalogs : 4004.0
Sports : 23008.898550724636
Music : 57326.530303030304
Photo & Video : 28441.54375
Travel : 28243.8


In [47]:
#print(ios_header)
for apps in ios_free_apps:
    genre = apps[11]
    
    if genre == "Reference":
        print(apps[1], ":", apps[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


For the Apple Store, References is a good option. Games and Social Networking is a very saturated genre and in-app purchases but not be the best option for these genres. 

In [46]:
android_categories = freq_table(android_free, 1)

for category in android_categories:
    total = 0
    len_category = 0
    
    for apps in android_free:
        category_app = apps[1]
        if category_app == category:
            installs = apps[5]
            installs = installs.replace("+", "")
            installs = installs.replace(",", "")
            installs = float(installs)
            total += installs
            len_category += 1
    avg_installs = total/len_category
    print(category, ":", avg_installs)

ART_AND_DESIGN : 1986335.0877192982
SPORTS : 3638640.1428571427
MAPS_AND_NAVIGATION : 4056941.7741935486
FINANCE : 1387692.475609756
BEAUTY : 513151.88679245283
COMICS : 817657.2727272727
FOOD_AND_DRINK : 1924897.7363636363
BUSINESS : 1712290.1474201474
TOOLS : 10801391.298666667
LIFESTYLE : 1437816.2687861272
ENTERTAINMENT : 11640705.88235294
TRAVEL_AND_LOCAL : 13984077.710144928
AUTO_AND_VEHICLES : 647317.8170731707
SOCIAL : 23253652.127118643
WEATHER : 5074486.197183099
BOOKS_AND_REFERENCE : 8767811.894736841
PRODUCTIVITY : 16787331.344927534
LIBRARIES_AND_DEMO : 638503.734939759
SHOPPING : 7036877.311557789
PARENTING : 542603.6206896552
VIDEO_PLAYERS : 24727872.452830188
DATING : 854028.8303030303
HEALTH_AND_FITNESS : 4188821.9853479853
FAMILY : 3695641.8198090694
EVENTS : 253542.22222222222
EDUCATION : 1833495.145631068
MEDICAL : 120550.61980830671
NEWS_AND_MAGAZINES : 9549178.467741935
PERSONALIZATION : 5201482.6122448975
COMMUNICATION : 38456119.167247385
GAME : 15588015.6032482

In [ ]:
for apps in android_free:
    if apps[0] == "Co"

In [ ]:
print(android_header)